### ISSUE TITLE
Implement custom parsing for quantile custom animation

### ISSUE THREAD

**Issue by jhofman** [microsoft/datamations#137]
now that @giorgi-ghviniashvili has implemented custom animations as per #18, we can extend the median animation to be for any quantile you want to compute, for instance the bottom 10% of the distribution, which is done by `quantile(x, 0.10)` in R.

@giorgi-ghviniashvili can implement this on the javascript side, he will expect a custom animation argument that gives the function and the quantile to be computed as a decimal, e.g.: `meta.custom_animation = "quantile(0.10)"`

then we'll implement it on the R side when ready.

**Comment by giorgi-ghviniashvili** [comment]
@jhofman as I said quantile was very easy to do. btw, `d3.median` is an alias to quantile(0.5). [see here](https://github.com/d3/d3-array/blob/main/src/median.js)

https://user-images.githubusercontent.com/6615532/150329137-4f97fd28-d292-4924-9ed2-47cb3cd0df24.mov

**Comment by giorgi-ghviniashvili** [comment]
Quantile animation fixed. 

https://user-images.githubusercontent.com/6615532/151383367-30d65fc0-b679-4676-8963-09312b8c0bf5.mov

**Comment by willdebras** [comment]
This is a pretty straightforward update I can commit tomorrow, specifically passing the meta.custom_animation the function and the `probs` value, e.g. quantile(0.1) as a string, but interested to know if instead we want this meta spec to have the listed results of a generic parsing function, like mentioned here:

https://github.com/microsoft/datamations/issues/18#issuecomment-1026451888

**Comment by willdebras** [comment]
Adding `spec[["meta"]][["custom_animation"]] <- tidy_function_args[["summarize"]][[2]]` to `prep_specs_summarize.R` generates a meta spec string like this.

**Comment by willdebras** [comment]
Have done some more debugging on this and have a good solution coming. Currently `datamation_sanddance()` fails to evaluate a data state for an expression like quantile(x, 0.1) because we aren't passing that `probs` value (0.1) in the prep_specs_summarize.R when we generate this data state.

Code like this:

```
"small_salary %>% 
  group_by(Degree) %>%
  summarise(quan = quantile(Salary, 0.5))" %>%
  datamation_sanddance()
```
Errors with something like this:

```
 Error: Problem with `mutate()` input `..1`.
i `..1 = dplyr::across(...)`.
i `..1` must be size 411 or 1, not 5.
i The error occurred in group 1: player = David Justice, year = 1995.
Run `rlang::last_error()` to see where the error occurred.
```

Specifically erroring on this line:
https://github.com/microsoft/datamations/blob/main/R/prep_specs_summarize.R#L476

I'm working on that generalized parsing right now to make sure this can evaluate data states for quantile animation. Planning to have a commit this weekend for review. This should give proper specs for @giorgi-ghviniashvili to compare against.

**Comment by jhofman** [comment]
got it, thanks for the update @willdebras!

**Comment by jhofman** [comment]
sounds like we will pass the unparsed summarize expression so it can get eval-ed exactly as given, and then have custom routines to pull out parameters that are relevant to animation (e.g., the 0.1 in `quantile(Salary, 0.1)`).

this will have to get propagated from `parse_functions` down through to `generate_specs`

**Comment by willdebras** [comment]
This is working now, with a new custom parameters syntax. We should be able to genericize it. Currently generating a new mapping that is a list of all parameters passed to the summary function and applying them when calling that summary_function in generating the data states.

https://github.com/microsoft/datamations/commit/c747cfec025bb4b8d593db60b91deb06bdf905e1

For @giorgi-ghviniashvili, I now have specs we can take a look at during our call:

https://github.com/microsoft/datamations/blob/custom_animations/sandbox/custom_animations/custom-animations-quantile-R.json

**Comment by willdebras** [comment]
The way these summary_parameters are handled is pretty nice. They can be named or unnamed in the original expression. The nature of this lets R do the work of interpreting what argument corresponds with what name.

```
"small_salary %>% 
  group_by(Degree) %>%
  summarise(quan = quantile(Salary, probs = 0.1))" %>%
  datamation_sanddance()
```
  and 
 ```
"small_salary %>% 
  group_by(Degree) %>%
  summarise(quan = quantile(Salary, 0.1))" %>%
  datamation_sanddance()
```

both return the following datamation.

https://user-images.githubusercontent.com/37971596/153608484-5939cf2f-1cba-49a2-9cd7-4037afb3cdaa.mp4

**Comment by jhofman** [comment]
awesome.

@giorgi-ghviniashvili, can you look at the new json spec that @willdebras generated and adjust your parsing so that you pick up the quantile parameter from the new custom_animation field?

**Comment by jhofman** [comment]
@willdebras, looks like the final frame specs might have the points reversed?

**Comment by jhofman** [comment]
@willdebras can you re-generate to check that this is all fine and then we can close the issue?

if good, then @giorgi-ghviniashvili can you send a pull request for custom animations so we can merge to main and test things out?

**Comment by willdebras** [comment]
@giorgi-ghviniashvili I confirmed these specs are up to date with specs generated by the main branch:

https://github.com/microsoft/datamations/blob/custom_animations/sandbox/custom_animations/custom-animations-quantile-R.json

If these run fine, I believe all custom animations should be good to merge. Once those are merged, I will make updates to the demo and documentation.

**Comment by giorgi-ghviniashvili** [comment]
This is quantile: 

https://user-images.githubusercontent.com/6615532/158153502-80d19674-917a-4ec3-8f11-e79a9c5370a2.mov

**Comment by jhofman** [comment]
complete and merged as of PR #144
